File: src/purra/artifacts/contracts.py
"""Storage-neutral contracts for recoverable model-generated artifacts."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any, Mapping

from purra.artifacts.ownership import ArtifactOwnerRef
from purra.normalization import (
    non_negative_int,
    optional_non_negative_int,
    optional_text,
    positive_int,
    required_text,
    text_tuple,
    unique_text_tuple,
)
from purra.json_values import (
    canonical_json_digest,
    freeze_json_mapping,
    thaw_json_mapping,
)


class ArtifactStatus(StrEnum):
    OPEN = "open"
    FINALIZED = "finalized"
    ABORTED = "aborted"


@dataclass(frozen=True, slots=True)
class ArtifactCreateCommand:
    namespace: str
    kind: str
    owner_id: str
    owner_ref: ArtifactOwnerRef
    created_by_run_id: str
    schema_version: int = 1
    expected_item_count: int | None = None
    metadata: Mapping[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        for name in ("namespace", "kind", "owner_id"):
            object.__setattr__(
                self,
                name,
                required_text(getattr(self, name), f"artifact {name}"),
            )
        if not isinstance(self.owner_ref, ArtifactOwnerRef):
            raise TypeError("artifact owner_ref must be an ArtifactOwnerRef")
        object.__setattr__(self, "created_by_run_id", required_text(
            self.created_by_run_id,
            "artifact created_by_run_id",
        ))
        object.__setattr__(
            self,
            "schema_version",
            positive_int(self.schema_version, "artifact schema_version"),
        )
        object.__setattr__(
            self,
            "expected_item_count",
            optional_non_negative_int(
                self.expected_item_count,
                "artifact expected_item_count",
            ),
        )
        object.__setattr__(self, "metadata", freeze_json_mapping(self.metadata))


@dataclass(frozen=True, slots=True)
class ArtifactRecord:
    id: str
    namespace: str
    kind: str
    owner_id: str
    owner_ref: ArtifactOwnerRef
    created_by_run_id: str
    schema_version: int
    status: ArtifactStatus = ArtifactStatus.OPEN
    revision: int = 1
    next_sequence: int = 1
    committed_item_count: int = 0
    expected_item_count: int | None = None
    metadata: Mapping[str, Any] = field(default_factory=dict)
    resource_ref: str | None = None
    coverage_digest: str | None = None

    def __post_init__(self) -> None:
        for name in ("id", "namespace", "kind", "owner_id"):
            object.__setattr__(
                self,
                name,
                required_text(getattr(self, name), f"artifact {name}"),
            )
        if not isinstance(self.owner_ref, ArtifactOwnerRef):
            raise TypeError("artifact owner_ref must be an ArtifactOwnerRef")
        object.__setattr__(self, "created_by_run_id", required_text(
            self.created_by_run_id,
            "artifact created_by_run_id",
        ))
        object.__setattr__(self, "status", ArtifactStatus(self.status))
        for name in ("schema_version", "revision", "next_sequence"):
            object.__setattr__(
                self,
                name,
                positive_int(getattr(self, name), f"artifact {name}"),
            )
        object.__setattr__(
            self,
            "committed_item_count",
            non_negative_int(
                self.committed_item_count,
                "committed_item_count",
            ),
        )
        object.__setattr__(
            self,
            "expected_item_count",
            optional_non_negative_int(
                self.expected_item_count,
                "artifact expected_item_count",
            ),
        )
        object.__setattr__(self, "metadata", freeze_json_mapping(self.metadata))
        object.__setattr__(
            self,
            "resource_ref",
            optional_text(self.resource_ref),
        )
        object.__setattr__(
            self,
            "coverage_digest",
            optional_text(self.coverage_digest),
        )

@dataclass(frozen=True, slots=True)
class ArtifactMutationLease:
    """Opaque proof that one Run currently owns an Artifact write.

    The repository, rather than a caller, decides whether the token is live and
    belongs to the Artifact.  ``lease_duration_ms`` is only a bounded renewal
    request applied atomically with a successful mutation.
    """

    run_id: str
    claim_token: str
    lease_duration_ms: int = 300_000

    def __post_init__(self) -> None:
        for name in ("run_id", "claim_token"):
            object.__setattr__(
                self,
                name,
                required_text(
                    getattr(self, name),
                    f"artifact mutation lease {name}",
                ),
            )
        object.__setattr__(
            self,
            "lease_duration_ms",
            positive_int(
                self.lease_duration_ms,
                "artifact mutation lease duration",
            ),
        )


@dataclass(frozen=True, slots=True)
class ArtifactAppendCommand:
    artifact_id: str
    expected_revision: int
    sequence: int
    batch_id: str
    idempotency_key: str
    items: tuple[Mapping[str, Any], ...]
    write_lease: ArtifactMutationLease
    coverage_keys: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        for name in ("artifact_id", "batch_id", "idempotency_key"):
            object.__setattr__(
                self,
                name,
                required_text(getattr(self, name), f"artifact batch {name}"),
            )
        for name in ("expected_revision", "sequence"):
            object.__setattr__(
                self,
                name,
                positive_int(getattr(self, name), f"artifact batch {name}"),
            )
        items = tuple(freeze_json_mapping(item) for item in self.items)
        if not items:
            raise ValueError("artifact batch must contain at least one item")
        object.__setattr__(self, "items", items)
        coverage = text_tuple(
            str(value or "").strip() for value in self.coverage_keys
        )
        object.__setattr__(self, "coverage_keys", coverage)
        if not isinstance(self.write_lease, ArtifactMutationLease):
            raise TypeError("artifact append write_lease is invalid")

    @property
    def content_digest(self) -> str:
        payload = {
            "artifactId": self.artifact_id,
            "sequence": self.sequence,
            "batchId": self.batch_id,
            "items": [thaw_json_mapping(item) for item in self.items],
            "coverageKeys": list(self.coverage_keys),
        }
        return canonical_json_digest(payload)


@dataclass(frozen=True, slots=True)
class ArtifactBatch:
    artifact_id: str
    batch_id: str
    idempotency_key: str
    sequence: int
    committed_revision: int
    items: tuple[Mapping[str, Any], ...]
    coverage_keys: tuple[str, ...] = ()
    content_digest: str = ""

    def __post_init__(self) -> None:
        for name in ("artifact_id", "batch_id", "idempotency_key"):
            object.__setattr__(
                self,
                name,
                required_text(getattr(self, name), f"artifact batch {name}"),
            )
        for name in ("sequence", "committed_revision"):
            object.__setattr__(
                self,
                name,
                positive_int(getattr(self, name), f"artifact batch {name}"),
            )
        object.__setattr__(
            self,
            "items",
            tuple(freeze_json_mapping(item) for item in self.items),
        )
        object.__setattr__(self, "coverage_keys", tuple(self.coverage_keys))
        object.__setattr__(
            self,
            "content_digest",
            required_text(
                self.content_digest,
                "artifact batch content_digest",
            ),
        )


@dataclass(frozen=True, slots=True)
class ArtifactBatchReceipt:
    artifact_id: str
    batch_id: str
    sequence: int
    committed_revision: int
    next_sequence: int
    accepted_count: int
    replayed: bool = False

    def __post_init__(self) -> None:
        for name in ("artifact_id", "batch_id"):
            object.__setattr__(
                self,
                name,
                required_text(getattr(self, name), f"artifact receipt {name}"),
            )
        for name in ("sequence", "committed_revision", "next_sequence"):
            object.__setattr__(
                self,
                name,
                positive_int(getattr(self, name), f"artifact receipt {name}"),
            )
        object.__setattr__(
            self,
            "accepted_count",
            non_negative_int(
                self.accepted_count,
                "artifact accepted_count",
            ),
        )
        object.__setattr__(self, "replayed", bool(self.replayed))


@dataclass(frozen=True, slots=True)
class ArtifactFinalizeCommand:
    artifact_id: str
    expected_revision: int
    write_lease: ArtifactMutationLease
    expected_item_count: int | None = None
    expected_coverage_keys: tuple[str, ...] = ()
    resource_ref: str | None = None

    def __post_init__(self) -> None:
        object.__setattr__(
            self,
            "artifact_id",
            required_text(self.artifact_id, "artifact_id"),
        )
        object.__setattr__(
            self,
            "expected_revision",
            positive_int(self.expected_revision, "expected_revision"),
        )
        object.__setattr__(
            self,
            "expected_item_count",
            optional_non_negative_int(
                self.expected_item_count,
                "artifact expected_item_count",
            ),
        )
        object.__setattr__(
            self,
            "expected_coverage_keys",
            unique_text_tuple(
                str(value or "").strip()
                for value in self.expected_coverage_keys
            ),
        )
        object.__setattr__(
            self,
            "resource_ref",
            optional_text(self.resource_ref),
        )
        if not isinstance(self.write_lease, ArtifactMutationLease):
            raise TypeError("artifact finalize write_lease is invalid")


@dataclass(frozen=True, slots=True)
class ArtifactValidationResult:
    accepted: bool
    code: str | None = None
    details: Mapping[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        object.__setattr__(self, "accepted", bool(self.accepted))
        code = optional_text(self.code)
        if not self.accepted and not code:
            raise ValueError("rejected artifact validation requires a code")
        object.__setattr__(self, "code", code)
        object.__setattr__(self, "details", freeze_json_mapping(self.details))


def coverage_digest(keys: tuple[str, ...]) -> str:
    return canonical_json_digest(sorted(set(keys)))


__all__ = [
    "ArtifactAppendCommand",
    "ArtifactBatch",
    "ArtifactBatchReceipt",
    "ArtifactCreateCommand",
    "ArtifactFinalizeCommand",
    "ArtifactMutationLease",
    "ArtifactRecord",
    "ArtifactStatus",
    "ArtifactValidationResult",
    "coverage_digest",
]

File: typescript/src/artifacts/contracts.ts
import type { JsonValue } from "../model/types.js";
import { copyJsonValue } from "../model/validation.js";
import { AgentError } from "../shared/errors.js";
import { stableFingerprint } from "../shared/fingerprint.js";
import type {
  ArtifactAccessRequest,
  ArtifactAppendCommand,
  ArtifactAppendOperation,
  ArtifactClaimLeaseCommand,
  ArtifactCreateCommand,
  ArtifactFinalizeCommand,
  ArtifactFinalizeOperation,
  ArtifactMaintenancePolicy,
  ArtifactMutationLease,
  ArtifactOwnerRef,
  ArtifactResumeCandidate,
  ArtifactValidationResult,
  ArtifactWriteClaimCommand,
} from "./types.js";

export function copyArtifactOwnerRef(value: ArtifactOwnerRef): ArtifactOwnerRef {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact ownerRef must be an object");
  }
  return Object.freeze({
    kind: requiredText(value.kind, "Artifact owner kind"),
    id: requiredText(value.id, "Artifact owner id"),
  });
}

export function copyArtifactCreateCommand(value: ArtifactCreateCommand): Required<ArtifactCreateCommand> {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact create command must be an object");
  }
  return Object.freeze({
    namespace: requiredText(value.namespace, "Artifact namespace"),
    kind: requiredText(value.kind, "Artifact kind"),
    ownerId: requiredText(value.ownerId, "Artifact ownerId"),
    ownerRef: copyArtifactOwnerRef(value.ownerRef),
    createdByRunId: requiredText(value.createdByRunId, "Artifact createdByRunId"),
    schemaVersion: positiveInteger(value.schemaVersion ?? 1, "Artifact schemaVersion"),
    expectedItemCount: nullableNonNegative(value.expectedItemCount ?? null, "Artifact expectedItemCount"),
    metadata: copyMapping(value.metadata ?? {}, "Artifact metadata"),
  });
}

export function copyArtifactMutationLease(
  value: ArtifactMutationLease,
): Required<ArtifactMutationLease> {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact mutation lease must be an object");
  }
  return Object.freeze({
    runId: requiredText(value.runId, "Artifact mutation Run id"),
    claimToken: requiredText(value.claimToken, "Artifact mutation claim token"),
    leaseDurationMs: positiveInteger(
      value.leaseDurationMs ?? 300_000,
      "Artifact mutation lease duration",
    ),
  });
}

export async function prepareArtifactAppend(
  value: ArtifactAppendCommand,
): Promise<ArtifactAppendOperation> {
  if (value === null || typeof value !== "object" || !Array.isArray(value.items)) {
    throw new TypeError("Artifact append command must contain items");
  }
  if (value.items.length === 0) throw new AgentError("artifact_batch_empty", "Artifact batch is empty");
  const artifactId = requiredText(value.artifactId, "Artifact id");
  const sequence = positiveInteger(value.sequence, "Artifact sequence");
  const batchId = requiredText(value.batchId, "Artifact batch id");
  const items = Object.freeze(value.items.map((item) => (
    copyMapping(item, "Artifact batch item")
  )));
  const coverageKeys = copyTextList(value.coverageKeys ?? [], "Artifact coverage key", false);
  const operation = {
    artifactId,
    expectedRevision: positiveInteger(value.expectedRevision, "Artifact expected revision"),
    sequence,
    batchId,
    idempotencyKey: requiredText(value.idempotencyKey, "Artifact idempotency key"),
    items,
    writeLease: copyArtifactMutationLease(value.writeLease),
    coverageKeys,
  };
  const contentDigest = await stableFingerprint(copyJsonValue({
    artifactId,
    sequence,
    batchId,
    items,
    coverageKeys,
  }));
  return Object.freeze({ ...operation, contentDigest });
}

export function copyArtifactFinalizeCommand(
  value: ArtifactFinalizeCommand,
): ArtifactFinalizeOperation {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact finalize command must be an object");
  }
  return Object.freeze({
    artifactId: requiredText(value.artifactId, "Artifact id"),
    expectedRevision: positiveInteger(value.expectedRevision, "Artifact expected revision"),
    writeLease: copyArtifactMutationLease(value.writeLease),
    expectedItemCount: nullableNonNegative(
      value.expectedItemCount ?? null,
      "Artifact expected item count",
    ),
    expectedCoverageKeys: copyTextList(
      value.expectedCoverageKeys ?? [],
      "Artifact expected coverage key",
      true,
    ),
    resourceRef: optionalText(value.resourceRef ?? null),
  });
}

export function copyArtifactResumeCandidate(
  value: ArtifactResumeCandidate,
): Required<ArtifactResumeCandidate> {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact resume candidate must be an object");
  }
  const status = artifactStatus(value.status);
  const committedItemCount = nonNegativeInteger(
    value.committedItemCount ?? 0,
    "Artifact committed item count",
  );
  const expectedItemCount = nullableNonNegative(
    value.expectedItemCount ?? null,
    "Artifact expected item count",
  );
  if (expectedItemCount !== null && committedItemCount > expectedItemCount) {
    throw new TypeError("Artifact committed item count exceeds its manifest");
  }
  return Object.freeze({
    artifactId: requiredText(value.artifactId, "Artifact candidate id"),
    namespace: requiredText(value.namespace, "Artifact candidate namespace"),
    kind: requiredText(value.kind, "Artifact candidate kind"),
    ownerId: requiredText(value.ownerId, "Artifact candidate ownerId"),
    ownerRef: copyArtifactOwnerRef(value.ownerRef),
    createdByRunId: requiredText(value.createdByRunId, "Artifact candidate createdByRunId"),
    status,
    revision: positiveInteger(value.revision, "Artifact candidate revision"),
    committedItemCount,
    expectedItemCount,
  });
}

export function copyArtifactAccessRequest(value: ArtifactAccessRequest): ArtifactAccessRequest {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact access request must be an object");
  }
  if (value.mode !== "read" && value.mode !== "write") {
    throw new TypeError("Artifact access mode must be read or write");
  }
  return Object.freeze({
    artifactId: requiredText(value.artifactId, "Artifact access id"),
    runId: requiredText(value.runId, "Artifact access Run id"),
    mode: value.mode,
    expectedRevision: positiveInteger(value.expectedRevision, "Artifact access revision"),
  });
}

export function copyArtifactWriteClaimCommand(
  value: ArtifactWriteClaimCommand,
): ArtifactWriteClaimCommand {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact write claim command must be an object");
  }
  return Object.freeze({
    artifactId: requiredText(value.artifactId, "Artifact claim id"),
    runId: requiredText(value.runId, "Artifact claim Run id"),
    expectedRevision: positiveInteger(value.expectedRevision, "Artifact claim revision"),
    leaseDurationMs: positiveInteger(value.leaseDurationMs, "Artifact claim lease duration"),
  });
}

export function copyArtifactClaimLeaseCommand(
  value: ArtifactClaimLeaseCommand,
): ArtifactClaimLeaseCommand {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact claim lease command must be an object");
  }
  return Object.freeze({
    artifactId: requiredText(value.artifactId, "Artifact claim id"),
    runId: requiredText(value.runId, "Artifact claim Run id"),
    claimToken: requiredText(value.claimToken, "Artifact claim token"),
    ...(value.leaseDurationMs === undefined
      ? {}
      : { leaseDurationMs: positiveInteger(value.leaseDurationMs, "Artifact claim lease duration") }),
  });
}

export function copyArtifactMaintenancePolicy(
  value: ArtifactMaintenancePolicy = {},
): Required<ArtifactMaintenancePolicy> {
  if (value === null || typeof value !== "object") {
    throw new TypeError("Artifact maintenance policy must be an object");
  }
  return Object.freeze({
    terminalRetentionMs: nullableNonNegative(
      value.terminalRetentionMs ?? null,
      "Artifact terminal retention",
    ),
    maxPurgeArtifacts: positiveInteger(
      value.maxPurgeArtifacts ?? 100,
      "Artifact maximum purge count",
    ),
  });
}

export function copyArtifactValidationResult(
  value: ArtifactValidationResult,
): Required<ArtifactValidationResult> {
  if (value === null || typeof value !== "object" || typeof value.accepted !== "boolean") {
    throw new TypeError("Artifact validator returned an invalid result");
  }
  const code = value.code === undefined ? "" : requiredText(value.code, "Artifact validation code");
  if (!value.accepted && code === "") {
    throw new TypeError("Rejected Artifact validation requires a code");
  }
  return Object.freeze({
    accepted: value.accepted,
    code,
    details: copyMapping(value.details ?? {}, "Artifact validation details"),
  });
}

export async function artifactCoverageDigest(keys: readonly string[]): Promise<string> {
  return stableFingerprint(Object.freeze([...new Set(keys)].sort()));
}

export function artifactOwnerKey(input: {
  readonly namespace: string;
  readonly kind: string;
  readonly ownerId: string;
  readonly ownerRef: ArtifactOwnerRef;
}): string {
  const ownerRef = copyArtifactOwnerRef(input.ownerRef);
  return [
    requiredText(input.namespace, "Artifact namespace"),
    requiredText(input.kind, "Artifact kind"),
    requiredText(input.ownerId, "Artifact ownerId"),
    ownerRef.kind,
    ownerRef.id,
  ].join("\u0000");
}

export function artifactStatus(value: unknown): "open" | "finalized" | "aborted" {
  if (value !== "open" && value !== "finalized" && value !== "aborted") {
    throw new TypeError("Artifact status is invalid");
  }
  return value;
}

export function copyMapping(
  value: Readonly<Record<string, JsonValue>>,
  label: string,
): Readonly<Record<string, JsonValue>> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${label} must be an object`);
  }
  return copyJsonValue(value) as Readonly<Record<string, JsonValue>>;
}

export function requiredText(value: unknown, label: string): string {
  if (typeof value !== "string" || value.trim() === "") throw new TypeError(`${label} is required`);
  return value.trim();
}

export function optionalText(value: unknown): string | null {
  if (value === undefined || value === null) return null;
  return requiredText(value, "Optional text");
}

export function positiveInteger(value: number, label: string): number {
  if (!Number.isSafeInteger(value) || value < 1) throw new TypeError(`${label} must be positive`);
  return value;
}

export function nonNegativeInteger(value: number, label: string): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new TypeError(`${label} must be non-negative`);
  return value;
}

function nullableNonNegative(value: number | null, label: string): number | null {
  return value === null ? null : nonNegativeInteger(value, label);
}

function copyTextList(
  value: readonly string[],
  label: string,
  unique: boolean,
): readonly string[] {
  if (!Array.isArray(value)) throw new TypeError(`${label}s must be an array`);
  const copied = value.map((item) => requiredText(item, label));
  if (unique && new Set(copied).size !== copied.length) {
    throw new TypeError(`${label}s must be unique`);
  }
  return Object.freeze(copied);
}
